src/sql/query.rs

```diff
@@ -1083,6 +1093,16 @@ fn plan_using_constraint(
                 column_name
             ),
         };
+        let l_type = &qcx.relation_type(&left).column_types[l];
+        let r_type = &qcx.relation_type(&right).column_types[r];
+        if l_type.scalar_type != r_type.scalar_type {
```

Comment:
@benesch The equivalent code for `ON a = b` checks for ScalarType::Unknown here. Can that happen in a USING join too? Will the types propagate down correctly in that case?

Reply:
I think it could if you join a table whose definition is `(SELECT NULL as col)`.

Reply:
Postgres:

```
jamie=# select * from (select 1 as a) as foo join (select null as a) as bar using (a);
ERROR:  JOIN/USING types integer and text cannot be matched
```

Does it infer the default type in the subquery before checking the outer query? I don't really understand how the unknown types work.

Reply:
```sql
benesch=# select * from (select 'a' as a) as foo join (select null as a) as bar using (a);
 a 
---
(0 rows)
```

Unknown just means we have a literal NULL and no additional context clues, like a type hint from a function call. Most commonly seen when NULL is used in a project list, like `SELECT NULL`. Postgres in some recent version added a pass that converts everything with unknown type in a project list to type text, and so the above works in Postgres.

With your patch as is:

```
materialize=> select * from (select 'a' as a) as foo join (select null as a) as bar using (a);
ERROR:  String and Unknown are not comparable (in NATURAL/USING join on a)
```

Since we don't presently perform that same unknown -> text rewrite it would be slightly more consistent to allow joining unknown columns against columns of any type. But we may want to follow Postgres's lead and add the rewrite at some point. The logic was that clients should never see an unknown type, and since clients are apprised of the types of all the expressions in the select list, those unknowns have to go—and so the unknown type is just something that exists transiently during planning.

Or something like that. My head hurts and I'm having trouble thinking.

Reply:
Do we get type unknown for query parameters too, or are they inferred separately?

Reply:
Also don't think too hard - none of this is that important. I can figure it out.

Reply:
Ha, if you let unknown join against text it breaks the new type-checking in coalesce:

```
message: coalesce inputs did not have uniform type: [ColumnType { nullable: false, scalar_type: String }, ColumnType { nullable: true, scalar_type: Unknown }]
```

I'll fix that too :)